src/file_io.rs

```diff
@@ -86,17 +108,57 @@ pub fn write_to_file(
     input: bio::io::fastq::Record,
     output: OutputFile,
     umi: &[u8],
-    edit_nr: bool,
+    umi_sep: Option<&String>,
+    edit_nr: Option<u8>,
 ) -> OutputFile {
     let s = input;
-    if edit_nr {
-        let header = &[s.id(), ":", std::str::from_utf8(&umi).unwrap()].concat();
+    let delim = umi_sep.as_ref().map(|s| s.as_str()).unwrap_or(":"); // the delimiter for the UMI
+    if let Some(number) = edit_nr {
+        let header = &[s.id(), delim, std::str::from_utf8(umi).unwrap()].concat();
         let mut string = String::from(s.desc().unwrap());
-        string.replace_range(0..1, "2");
+        string.replace_range(0..1, &number.to_string());
         let desc: Option<&str> = Some(&string);
         output.write(header, desc, s)
     } else {
-        let header = &[s.id(), ":", std::str::from_utf8(&umi).unwrap()].concat();
+        let header = &[s.id(), delim, std::str::from_utf8(umi).unwrap()].concat();
         output.write(header, s.desc(), s.clone())
     }
 }
+
+// Checks whether an output path exists.
+pub fn check_outputpath(mut path: PathBuf, compress: &bool, force: &bool) -> Result<PathBuf> {
+    // handle the compression and adapt file extension if necessary.
+    if let Some(extension) = path.extension().and_then(|e| e.to_str()) {
+        if !extension.ends_with("gz") & compress {
+            let mut new_extension = extension.to_owned();
+            new_extension.push_str(".gz");
+            path.set_extension(new_extension);
+        }
+    }
+
+    // check if the path already exists
+    let exists = fs::metadata(&path).is_ok();
+
+    // return the path of it is ok to write, otherwise an error.
+    if exists & !force {
+        // force will disable prompt, but not the check.
+        if Confirm::new()
+            .with_prompt(format!("{} exists. Overwrite?", path.display()))
+            .interact()?
+        {
+            println!("File will be overwritten.");
+            Ok(path)
+        } else {
+            Err(anyhow!(RuntimeErrors::FileExistsError))
+        }
+    } else {
+        Ok(path)
+    }
+}
+
+pub fn append_umi_to_path(path: &Path) -> PathBuf {
```

Comment:
It is, but this function does just auto-generates the output file name if no explicit one has been specified.

The extension is handled by `check_outputpath()`, where indeed `compress` is used to set a `.gz` suffix, if missing:

```Rust
 if !extension.ends_with("gz") & compress {
            let mut new_extension = extension.to_owned();
            new_extension.push_str(".gz");
            path.set_extension(new_extension);
        }
```

Reply:
My rationale was, that this level of control is enabled by explicitly specifying the output file name, which will surely be done in e.g. a Nextflow pipeline. If `--out` and `--out2` are omitted, I believe, that _**output like input**_ is most safe to assume. 

If you favour _**uncompressed output unless explicitly specified**_, the easiest is to change the condition to `if extension == "gz" & compress` here:
```// Create output files
pub fn output_file(name: PathBuf) -> OutputFile {
    if let Some(extension) = name.extension() {
        if extension == "gz" {
            // File has gz extension, which has been enforced by check_outputpath() if -z was provided.
            OutputFile::Gzip {
                read: std::fs::File::create(name.as_path())
                    .map(|w| flate2::write::GzEncoder::new(w, flate2::Compression::default()))
                    .map(bio::io::fastq::Writer::new)
                    .unwrap(),
            }
        } else {...}
```

This also disables implicit compression, when people just add the extension to the output file name - something I considered a nice feature, but which might cause issues in conjunction with FIFOs.

Ultimately, the whole disagreement comes down to the poor compression performance of `umi-transfer` itself. Generally, people will likely never want to work with uncompressed output, if not for using FIFOs to compress it to improve performance. Given that we have refactored basically everything now, we might as well look at multithreaded compression still - it can't be much more complicated, can it?

Reply:
Yeah it's not a bad behaviour actually, but it's a bit hard to guess from a user perspecive, but I'm ok with it if it's intended and documented. First time I reported it I thought it was unintended. And I'm still a tiny bit bugged by the `--compress` flag being useless for that niche scenario.

In principle I agree that efficient compression should be included in umi-transfer, but I'd say we have to wait with that until later.